File: contrib/jsimplemodule/example/Source.java
import org.omnetpp.simkernel.*;

/**
 * Sends a message periodically.
 */
public class Source extends JSimpleModule {
    private int counter = 0;

    protected void initialize() {
        ev.println("initialize of "+getFullPath());
        cMessage msg = new cMessage("timeout-"+getFullName());
        scheduleAt(simTime() + 1, msg);
    }

    protected void handleMessage(cMessage msg) {
        String msgname = "msg-" + counter++;
        ev.println("sending "+msgname);
        send(new cMessage(msgname), "out");

        ev.println("scheduling next send");
        scheduleAt(simTime() + Simkernel.exponential(1), msg);
    }

    protected void finish() {
        ev.println("finish of "+getFullPath());
    }
};

File: ui/org.omnetpp.cdt/src/org/omnetpp/cdt/wizard/templates/Source.cc
{{copyright}}

#include "Source.h"

Define_Module(Source);

Source::Source()
{
    timerMessage = NULL;
}

Source::~Source()
{
    cancelAndDelete(timerMessage);
}

void Source::initialize()
{
    timerMessage = new cMessage("timer");
    scheduleAt(simTime(), timerMessage);
}

void Source::handleMessage(cMessage *msg)
{
    ASSERT(msg==timerMessage);

    cMessage *job = new cMessage("job");
    send(job, "out");

    scheduleAt(simTime()+par("sendIaTime").doubleValue(), timerMessage);
}

